### PR TITLE
XWIKI-15666: Validation check broken for 2 consecutive line breaks on translation pages

### PR DESCRIPTION
**Pull request by surli**
## Issue

https://jira.xwiki.org/browse/XWIKI-15666

## Solution

As discussed in https://markmail.org/message/okihfnpkavhnbkib the proposed solution for now is to skip technical pages from WCAG validator.
This has been done by checking a boolean `skipTechnicalPages` in the validator classes: if set to true, then the pages are parsed to get the value of the `hidden` attribute.

### CODE REVIEW

File: xwiki-platform-distribution/xwiki-platform-distribution-flavor/xwiki-platform-distribution-flavor-test/xwiki-platform-distribution-flavor-test-webstandards/src/test/it/org/xwiki/test/webstandards/framework/AbstractValidationTest.java [hunk 1]
```diff
@@ -217,6 +218,11 @@ protected static void addXarFiles(Class< ? extends AbstractValidationTest> valid
         }
     }
 
+    private static boolean isNotTechnicalPage(XarEntry entry)
+    {
+        return DefaultXarEntryType.HINT.equals(entry.getEntryType());
```
Comment (tmortagne) [inline xwiki-platform-distribution/xwiki-platform-distribution-flavor/xwiki-platform-distribution-flavor-test/xwiki-platform-distribution-flavor-test-webstandards/src/test/it/org/xwiki/test/webstandards/framework/AbstractValidationTest.java]: This is a very different logic. Default type is used to protect pages which are not supposed to be modified by users which is very different from pages user are not supposed to see.

**Comment by surli** [comment]
This is a first proposal for technical page skipping: I'm not sure about this solution for two reasons:
  1. it relies on the XarEntryType to only keep the "default" pages: from what I understood by reading the javadoc, the other types seems quite close to the technical page we want to avoid, but I'm not sure it's 100% accurate. 

  2. here it skips technical pages in all validators, since this filter is statically called. I might add a static flag to allow or not technical pages skipping in the right validator, but it looks not really good.

File: xwiki-platform-distribution/xwiki-platform-distribution-flavor/xwiki-platform-distribution-flavor-test/xwiki-platform-distribution-flavor-test-webstandards/src/test/it/org/xwiki/test/webstandards/framework/AbstractValidationTest.java (hunk 1 above)
Comment (surli) [reply xwiki-platform-distribution/xwiki-platform-distribution-flavor/xwiki-platform-distribution-flavor-test/xwiki-platform-distribution-flavor-test-webstandards/src/test/it/org/xwiki/test/webstandards/framework/AbstractValidationTest.java]: OK. I wasn't quite sure about that :)

File: xwiki-platform-core/xwiki-platform-xar/xwiki-platform-xar-model/src/main/java/org/xwiki/xar/XarPackage.java [hunk 2]
```diff
@@ -231,6 +237,35 @@ public void read(ZipFile zipFile) throws IOException, XarException
...
+     * {@link XarFile#getInputStream(LocalDocumentReference)}, else a FileInputStream is created from the file
+     * corresponding to the document reference if it exists.
+     *
+     * @param localDocumentReference the reference for which to return the InputStream.
+     * @return an InputStream corresponding to the given reference.
+     * @throws IOException in case the entry or the file cannot be found.
+     * @throws IOException in case of errors when creating the {@link XarFile}.
+     * @since 10.11RC1
+     */
+    public InputStream getInputStream(LocalDocumentReference localDocumentReference) throws IOException, XarException
```
Comment (tmortagne) [inline xwiki-platform-core/xwiki-platform-xar/xwiki-platform-xar-model/src/main/java/org/xwiki/xar/XarPackage.java]: I really don't like having this kind of API in XarPackage, it's not supposed to be its job.

If you really need a XarFile equivalent for folders then you should introduce one IMO.

Comment (surli) [reply xwiki-platform-core/xwiki-platform-xar/xwiki-platform-xar-model/src/main/java/org/xwiki/xar/XarPackage.java]: > If you really need a XarFile equivalent for folders then you should introduce one IMO.

Actually it's linked to the process of xwiki-platform-distribution-flavor-test-webstandards which I don't fully understand: during the preparation before the test, pages are exported in a folder, but without packaging it into a XAR, and then the `XarPackage.getEntries()` is used to read them.

So maybe we should just package the XAR?

File: xwiki-platform-distribution/xwiki-platform-distribution-flavor/xwiki-platform-distribution-flavor-test/xwiki-platform-distribution-flavor-test-webstandards/src/test/it/org/xwiki/test/webstandards/framework/AbstractValidationTest.java [hunk 3]
```diff
@@ -211,24 +219,75 @@ protected static void addXarFiles(Class< ? extends AbstractValidationTest> valid
...
+        for (DocumentReference documentReference : readXarContents(path, patternFilter, skipTechnicalPages)) {
             suite.addTest(validationTest.getConstructor(Target.class, HttpClient.class, Validator.class, String.class)
                 .newInstance(new DocumentReferenceTarget(documentReference), client, validator, "Admin:admin"));
         }
     }
 
+    private static boolean isTechnicalPage(String directoryPath, XarEntry entry, DocumentBuilder documentBuilder)
+        throws Exception
+    {
+        InputStream inputStream = new FileInputStream(new File(directoryPath, entry.getEntryName()));
```
Comment (tmortagne) [inline xwiki-platform-distribution/xwiki-platform-distribution-flavor/xwiki-platform-distribution-flavor-test/xwiki-platform-distribution-flavor-test-webstandards/src/test/it/org/xwiki/test/webstandards/framework/AbstractValidationTest.java]: You don't see to close this stream anywhere.

**Comment by surli** [comment]
fixed.

File: xwiki-platform-distribution/xwiki-platform-distribution-flavor/xwiki-platform-distribution-flavor-test/xwiki-platform-distribution-flavor-test-webstandards/src/test/it/org/xwiki/test/webstandards/framework/AbstractValidationTest.java [hunk 4]
```diff
@@ -211,24 +219,76 @@ protected static void addXarFiles(Class< ? extends AbstractValidationTest> valid
...
+    {
+        try (InputStream inputStream = new FileInputStream(new File(directoryPath, entry.getEntryName()))) {
+            Document parsedDocument = documentBuilder.parse(inputStream);
+
+            NodeList elements = parsedDocument.getElementsByTagName("hidden");
+
+            boolean isHiddenPage = false;
+            if (elements.getLength() == 1) {
+                    isHiddenPage = "true".equals(elements.item(0).getTextContent());
+                }
```
Comment (tmortagne) [inline xwiki-platform-distribution/xwiki-platform-distribution-flavor/xwiki-platform-distribution-flavor-test/xwiki-platform-distribution-flavor-test-webstandards/src/test/it/org/xwiki/test/webstandards/framework/AbstractValidationTest.java]: Surprising indentation :)

**Comment by tmortagne** [comment]
Looks OK but as reported by @mflorea in https://markmail.org/message/okihfnpkavhnbkib we might want to still enable some of the hidden pages. Now that you parse the document would be nice to also add at least a class white list. Link input is probably enough for other use cases.

**Comment by surli** [comment]
> Link input is probably enough for other use cases.

not sure what you mean here. But ok I'll add the whitelist to check against className of the document.

**Comment by tmortagne** [comment]
I mean that sure we don't check hidden documents by default but among those hidden documents we would still like to validate documents containing an object with a class which is part of a configured white list (on example being `Panels.PanelClass` because panels are visible stuff we want to validate but they are only visible when selected).

**Comment by surli** [comment]
Just to be sure: if there is at least one object using a class whitelisted, then the whole document should be checked, right?

**Pull request merged**